Arrow Database Compression

abhijithanilkumar opened the pull request:
* TPL built-ins for the following use cases:
    * To copy a tuple from one location to another
    * To insert into a specific TupleSlot
    * To allocate a new TupleSlot 
* New test in `sample_tpl/` that demonstrates the working of the newly created built-ins which inserts into a specific slot along with index updates.
* Added a new test file: `block_compactor_tpl_test.cpp` to verify the working of the newly added TPL built-in.
The original behaviour of `BlockCompactor` is retained, and we added a new function, `MoveTupleTPL` to demonstrate the working of the newly added TPL built-ins. For now, we have used hard-coded TPL code inside the `BlockCompactor`, this needs to be changed to generate code via `CodeGen` once the dependency issue is resolved. As mentioned in Future Work, it should be redesigned to prevent the circular dependency to `catalog` so that an execution context can be constructed from within the `BlockCompactor`.


You can find the design doc for the Project [here](https://github.com/721-Project-Team/Design-Docs).

xuanxuange commented on test/storage/data_table_test.cpp:
```diff
@@ -321,4 +335,33 @@ TEST_F(DataTableTests, InsertNoWrap) {
+
+  // Once the tuple is inserted, the insertion head should be incremented by 1
+  EXPECT_EQ(block->GetInsertHead(), 1);
+
```
Maybe the test can also check the size of inseted_tuples and the actual tuple content being inserted.

abhijithanilkumar replied:
We decided to move away from this design, marking as resolved.

DUKaige commented on src/storage/data_table.cpp:
```diff
@@ -210,6 +210,55 @@ TupleSlot DataTable::Insert(const common::ManagedPointer<transaction::Transactio
+  return block;
+}
+
+bool DataTable::InsertIntoFreezingBlock(common::ManagedPointer<transaction::TransactionContext> txn,
```
Do blocks under FREEZING status all have valid arrow headers? If so, should you also update the num_records in the header? I might have missed something.

abhijithanilkumar replied:
We decided to move away from this design, marking as resolved.

tpan496 commented on src/include/storage/sql_table.h:
```diff
@@ -110,6 +110,23 @@ class SqlTable {
+                               const ProjectedRow &redo, RawBlock *block) {
+    TERRIER_ASSERT(
+        block->controller_.GetBlockState()->load() == BlockState::FREEZING,
+        "InsertIntoFreezingBlock should be called only from the compactor, and the block should be FREEZING.");
```
How do you plan to enforce that this function can be only called from the compactor?

abhijithanilkumar replied:
We decided to move away from this design, marking as resolved.

abhijithanilkumar commented:
Code Review #2 Update:

* Added new built-ins to be used in the Block Compactor
* Added a new test in `sample_tpl` to test the new builtins
* Compiling the TPL code in the BlockCompactor constructor bypassing CodeGen (for now)

xuanxuange commented on test/storage/block_compactor_test.cpp:
```diff
@@ -1,390 +1,240 @@
+  transaction::TransactionManager *txn_manager_ptr = txn_manager_.Get();
+  compactor.ProcessCompactionQueue(deferred_action_manager_ptr, txn_manager_ptr);
+
+  // Check for correctness of compaction
```
I think it might be good to also check the arrow metadata after compaction so that the num_records, col_info are all correct.

abhijithanilkumar replied:
We are not using this test as of now, marking as resolved.

xuanxuange commented on sample_tpl/tpl_tests.txt:
```diff
@@ -56,6 +56,7 @@ agg-vec-filter.tpl,true,10
 static-agg.tpl,true,49995000
 distinct-agg.tpl,true,45
 delete.tpl,true,0
+compaction-insert-into.tpl,true,11
```
Is there any way to test if the tuple is inserted into the index successfully? either here or in the block_compactor test. I think select works even without index in our part. (correct me if that's incorrect)

abhijithanilkumar replied:
The test also checks index inserts and it returns an error in case it fails.

xuanxuange commented on test/storage/block_compactor_test.cpp:
```diff
@@ -1,390 +1,240 @@
+#include "transaction/transaction_manager.h"
 
-#define EXPORT_TABLE_NAME "test_table.arrow"
+#define LOG_FILE_NAME "./test.log"
```
I guess your test don't need logging actually.

xuanxuange commented on test/storage/block_compactor_test.cpp:
```diff
@@ -1,390 +1,240 @@
+      MakeExecCtx(db_oid, txn3, common::ManagedPointer(
+          catalog_->GetAccessor(common::ManagedPointer(txn3),
+              db_oid)
+      )); // Possible error -> Two pointers arent set (last 2 arguments not passed)
```
Fix this??

abhijithanilkumar replied:
Resolved.

abhijithanilkumar commented on src/include/storage/data_table.h:
```diff
@@ -200,6 +206,35 @@ class DataTable {
+   * @param block the block to which the tuple should be inserted into
+   * @return true if successful, false otherwise
+   */
+  bool InsertIntoFreezingBlock(common::ManagedPointer<transaction::TransactionContext> txn, const ProjectedRow &redo,
```
You are right, this function is not required anymore. We'll remove it in the next commit.

abhijithanilkumar replied:
Resolved.

DUKaige commented on src/include/storage/block_compactor.h:
```diff
@@ -49,6 +54,36 @@ class BlockCompactor {
+    //      cg->table_->InsertInto(common::ManagedPointer(cg->txn_), *record->Delta(), to);
+    //      return cg->table_->Delete(common::ManagedPointer(cg->txn_), from);
+    // @todo: do we need to consider that insertIndex could fail? Do so now.
+    auto tpl_code = R"(
```
Might be better if the tpl code is defined as a constant in the beginning of the file..

DUKaige replied:
Is there any better alternative to hardcoding tpl code as a string and compile at runtime?

abhijithanilkumar replied:
Yes, this needs to be done within CodeGen. But the downside is that we can only do that once the circular dependency with catalog is resolved. We have mentioned this in our design doc.

DUKaige commented on src/include/storage/block_compactor.h:
```diff
@@ -49,6 +54,36 @@ class BlockCompactor {
   };
 
  public:
+  BlockCompactor(execution::exec::ExecutionContext *exec, col_id_t *col_oids, const char *table_name) {
```
Some documentation on this constructor can be helpful.

abhijithanilkumar replied:
Resolved.

DUKaige commented on src/include/storage/block_compactor.h:
```diff
@@ -49,6 +54,36 @@ class BlockCompactor {
+    auto tpl_code = R"(
+    fun moveTuple(execCtx: *ExecutionContext, slot_from: *TupleSlot, slot_to: *TupleSlot, col_oids: *uint16, table_name: *uint8) -> bool {
+      var storage_interface: StorageInterface
+      //@storageInterfaceInitBind(&storage_interface, execCtx, table_name, col_oids, true)
```
Dead code

abhijithanilkumar replied:
Resolved

abhijithanilkumar commented on src/storage/data_table.cpp:
```diff
@@ -210,6 +210,55 @@ TupleSlot DataTable::Insert(const common::ManagedPointer<transaction::Transactio
+  bool status = false;
+  TupleSlot new_slot;
+  // Set block status to busy
+  accessor_.SetBlockBusyStatus(block);
```
We decided to move away from this design, marking as resolved.

abhijithanilkumar commented on src/include/storage/data_table.h:
```diff
@@ -200,6 +200,25 @@ class DataTable {
    */
   TupleSlot Insert(common::ManagedPointer<transaction::TransactionContext> txn, const ProjectedRow &redo);
 
+  /**
```
We decided to move away from this design, marking as resolved.

pdaviso commented:
The final code has been dropped! It meets all the requirements specified. You can find the design doc for the Project [here](https://github.com/721-Project-Team/Design-Docs).

apavlo commented:
Closing to reduce Jenkins contention
